README.md
# THIS PROJECT ISN'T MAINTAINED ANYMORE
## PLEASE UPGRADE TO THE OFFICAL GEM [mollie-api-ruby](https://github.com/mollie/mollie-api-ruby)


CHANGELOG.md
# Changelog

- **22 August 2014**: version 1.0.5
    - Final release with deprication message

- **2 July 2013**: version 1.0.4
    - Rails 4 support

- **25 June 2013**: version 1.0.3
    - Make Nokogiri version less restrictive by [@tijmenb](https://github.com/tijmenb)
    - Added different argument passing for `IdealMollie.new_order`

- **11 Februari 2012**: version 1.0.2
    - Fix for verify sha512

- **10 July 2012**: version 1.0.0
    - Released version 1.0
    - Added more platforms to test on Travis CI
    - Fixed spec for latest version of RSpec (2.11.0)

- **9 May 2012**: version 0.0.8
	- Upgraded dependency multi_xml to 0.5

- **26 April 2012**: version 0.0.7
    - Upgraded dependency faraday to 0.8
    - Upgraded dependency vcr to 2.1
    - Recorded the vcr cassettes to match the 2.1 format
	- Removed jruby dependency

- **30 January 2012**: version 0.0.6
    - Changed the URL to a CONST for testing on local sinatra server (coming soon)

- **30 January 2012**: version 0.0.5
    - Added support for multiple profile's

- **26 January 2012**: version 0.0.4
    - Changed property `payed` to `paid`

- **26 January 2012**: version 0.0.3
    - Added `Bank`, `Order` and `OrderResult`. Should be easier then hashes.
    - Added example to the README
    - Now you can override the return_url within the controller

- **25 January 2012**: version 0.0.2
    - Dependencies fixed
    - Tests fixed
    - Automation with Travis and Gemnasium

- **25 January 2012**: version 0.0.1
    - Initial release


spec/config_spec.rb
require 'spec_helper'

describe IdealMollie::Config do
  before(:each) do
    @config = IdealMollie::Config
    @config.test_mode = true
    @config.partner_id = 123456
    @config.profile_key = "321zyx12"
    @config.report_url = "http://example.org/report"
    @config.return_url = "http://example.org/return"
  end

  describe "#reset!" do
    it "should reset the values" do
      @config.reset!
      @config.test_mode.should be_false
      @config.partner_id.should be_nil
      @config.profile_key.should be_nil
      @config.report_url.should be_nil
      @config.return_url.should be_nil
    end
  end

  describe "#update!" do
    it "should update" do
      @config.reset!
      @config.test_mode = false
      @config.partner_id = 987654
      @config.profile_key = "123abc45"
      @config.report_url = "report"
      @config.return_url = "return"
      @config.update!

      config = IdealMollie::Config
      config.test_mode.should be_false
      config.partner_id.should eq 987654
      config.profile_key.should eq "123abc45"
      config.report_url.should eq "report"
      config.return_url.should eq "return"
    end
  end
end


spec/spec_helper.rb
require 'simplecov'
require 'coveralls'
SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter[
  SimpleCov::Formatter::HTMLFormatter,
  Coveralls::SimpleCov::Formatter
]
SimpleCov.start

require 'rubygems'
require 'bundler/setup'
require 'vcr'
require 'cgi'
require 'ideal-mollie'

VCR.configure do |c|
  c.cassette_library_dir = 'spec/vcr_cassettes'
  c.hook_into :fakeweb
  c.register_request_matcher :ignore_query_param_ordering do |r1, r2|
    uri1 = URI(r1.uri)
    uri2 = URI(r2.uri)

    uri1.scheme == uri2.scheme &&
      uri1.host == uri2.host &&
        uri1.path == uri2.path &&
          CGI.parse(uri1.query) == CGI.parse(uri2.query)
  end
end

RSpec.configure


spec/ideal_mollie_spec.rb
require 'spec_helper'

describe IdealMollie do
  before(:each) do
    @config = IdealMollie::Config
    @config.reset!
    @config.test_mode = false
    @config.partner_id = 987654
    @config.report_url = "http://example.org/report"
    @config.return_url = "http://example.org/return"
  end

  context "#banks" do
    it "returns an array with banks" do
      VCR.use_cassette("banks", :match_requests_on => [:ignore_query_param_ordering]) do
        banks = IdealMollie.banks
        banks.is_a?(Array).should be_true
        banks.count > 0

        bank = banks.first
        bank.id.should eq "0031"
        bank.name.should eq "ABN AMRO"
      end
    end
  end

  context "#new_order" do
    it "should return a Order with the correct values" do
      VCR.use_cassette("new_order", :match_requests_on => [:ignore_query_param_ordering]) do
        order = IdealMollie.new_order(1000, "test", "0031")
        order.transaction_id.should eq "c9f93e5c2bd6c1e7c5bee5c5580c6f83"
        order.amount.should eq 1000
        order.currency.should eq "EUR"
        order.url.should eq "https://www.abnamro.nl/nl/ideal/identification.do?randomizedstring=8433910909&trxid=30000217841224"
        order.message.should eq "Your iDEAL-payment has successfully been setup. Your customer should visit the given URL to make the payment"
      end
    end
    it "should return a Order for profile_key with the correct values" do
      @config.profile_key = "123abc45"
      @config.update!
      VCR.use_cassette("new_order", :match_requests_on => [:ignore_query_param_ordering]) do
        order = IdealMollie.new_order(1000, "test", "0031")
        order.transaction_id.should eq "474ed7b2735cbe4d1f4fd4da23269263"
        order.amount.should eq 1000
        order.currency.should eq "EUR"
        order.url.should eq "https://www.abnamro.nl/nl/ideal/identification.do?randomizedstring=6616737002&trxid=30000226032385"
        order.message.should eq "Your iDEAL-payment has successfully been setup. Your customer should visit the given URL to make the payment"
      end
    end
    it "should override the return url when specified" do
      params = IdealMollie.new_order_params(1200, "test", "0031")
      params[:returnurl].should eq "http://example.org/return"

      params = IdealMollie.new_order_params(1200, "test", "0031", "http://another.example.org/return")
      params[:returnurl].should eq "http://another.example.org/return"
    end
    it "should not append the profile_key this isn't specified in the config" do
      params = IdealMollie.new_order_params(1200, "test", "0031")
      params.has_key?(:profile_key).should be_false
    end
    it "should append the profile_key if specified in the config" do
      @config.profile_key = 12345
      @config.update!

      params = IdealMollie.new_order_params(1200, "test", "0031")
      params.has_key?(:profile_key).should be_true
    end
    it "should override the report url when specified" do
      params = IdealMollie.new_order_params(1200, "test", "0031")
      params[:reporturl].should eq "http://example.org/report"

      params = IdealMollie.new_order_params(1200, "test", "0031", nil, "http://another.example.org/report")
      params[:reporturl].should eq "http://another.example.org/report"
    end
    it "should accept hash as arguments for new_order" do
      VCR.use_cassette("new_order", :match_requests_on => [:ignore_query_param_ordering]) do
        order = IdealMollie.new_order(amount: 1000, description: "test", bank_id: "0031")
        order.amount.should eq 1000
      end
    end
    it "should accept changing the report_url in the hash for new_order" do
      VCR.use_cassette("new_order", :match_requests_on => [:ignore_query_param_ordering]) do
        # should recieve
        params = {
          :partnerid => 987654,
          :reporturl => "http://another.example.org/report",
          :returnurl => "http://example.org/return",
          :description => "test",
          :amount => 1000,
          :bank_id => "0031"
        }
        # dummy result
        result = {}
        result["order"] = nil

        IdealMollie.should_receive(:request).with("fetch", params).and_return(result)

        order = IdealMollie.new_order(
          amount: 1000,
          description: "test",
          bank_id: "0031",
          report_url: "http://another.example.org/report")
      end
    end
    it "should accept changing the return_url in the hash for new_order" do
      VCR.use_cassette("new_order", :match_requests_on => [:ignore_query_param_ordering]) do
        # should recieve
        params = {
          :partnerid => 987654,
          :reporturl => "http://example.org/report",
          :returnurl => "http://another.example.org/return",
          :description => "test",
          :amount => 1000,
          :bank_id => "0031"
        }
        # dummy result
        result = {}
        result["order"] = nil

        IdealMollie.should_receive(:request).with("fetch", params).and_return(result)

        order = IdealMollie.new_order(
          amount: 1000,
          description: "test",
          bank_id: "0031",
          return_url: "http://another.example.org/return")
      end
    end
  end

  context "#check_order" do
    it "should return a OrderResult with the correct values" do
      VCR.use_cassette("check_order", :match_requests_on => [:ignore_query_param_ordering]) do
        order_result = IdealMollie.check_order("c9f93e5c2bd6c1e7c5bee5c5580c6f83")
        order_result.transaction_id.should eq "c9f93e5c2bd6c1e7c5bee5c5580c6f83"
        order_result.amount.should eq 1000
        order_result.currency.should eq "EUR"
        order_result.paid.should eq false
        order_result.message.should eq "This iDEAL-order wasn't payed for, or was already checked by you. (We give payed=true only once, for your protection)"
        order_result.status.should eq "CheckedBefore"
      end
    end

    it "should mark the OrderResult as paid and contain the customer information when called by mollie" do
      VCR.use_cassette("check_order", :match_requests_on => [:ignore_query_param_ordering]) do
        order_result = IdealMollie.check_order("482d599bbcc7795727650330ad65fe9b")
        order_result.transaction_id.should eq "482d599bbcc7795727650330ad65fe9b"
        order_result.amount.should eq 1000
        order_result.currency.should eq "EUR"
        order_result.paid.should eq true
        order_result.message.should eq "This iDEAL-order has successfuly been payed for, and this is the first time you check it."

        order_result.customer_name.should eq "Hr J Janssen"
        order_result.customer_account.should eq "P001234567"
        order_result.customer_city.should eq "Amsterdam"
      end
    end
  end
end


lib/ideal-mollie/order.rb
module IdealMollie
  #
  # Object representing a "Order" object with attributes provided by Mollie
  #
  # @example
  #   Order.new({
  #     :transaction_id => "c9f93e5c2bd6c1e7c5bee5c5580c6f83",
  #     :amount => "1000",
  #     :currency => "EUR",
  #     :url => "https://www.abnamro.nl/nl/ideal/identification.do?randomizedstring=8433910909&trxid=30000217841224",
  #     :message => "Your iDEAL-payment has successfully been setup. Your customer should visit the given URL to make the payment"
  #   })
  #
  # @note The amount is always specified in *cents*. So 10,00 EUR would be 1000.
  class Order
    # @return [String] A unique id generated by Mollie.
    attr_accessor :transaction_id
    # @return [int] The amount of money for the transaction.
    attr_accessor :amount
    # @return [String] The currency for the transaction.
    attr_accessor :currency
    # @return [String] The url to the bank, to complete payment.
    attr_accessor :url
    # @return [String] A message with additional information about the Order.
    attr_accessor :message

    #
    # Initializer to transform a +Hash+ into an Order object
    #
    # @param [Hash] values
    def initialize(values=nil)
      return if values.nil?

      @transaction_id = values["transaction_id"].to_s if values.has_key?("transaction_id")
      @amount = values["amount"].to_i if values.has_key?("amount")
      @currency = values["currency"].to_s if values.has_key?("currency")
      @url = values["URL"].to_s if values.has_key?("URL")
      @message = values["message"].to_s if values.has_key?("message")
    end
  end
end


lib/ideal-mollie/config.rb
#
# Configuration object for storing some parameters required for making transactions
#
module IdealMollie::Config
  class << self
    # @return [int] Your Mollie partner id.
    # @note See: https://www.mollie.nl/beheer/betaaldiensten/documentatie/ideal/ for your partner id
    attr_accessor :partner_id
    # @return [String] Your Mollie profile key
    # @note The is a optional parameter. You only need this if you have multiple profiles
    # @note See: https://www.mollie.nl/beheer/betaaldiensten/profielen/ for the list of profiles
    attr_accessor :profile_key
    # @return [String] The url Mollie uses to report the status of the payment
    attr_accessor :report_url
    # @return [String] The url Mollie sends you to when a transaction is finished
    attr_accessor :return_url
    # @return [Boolean] Test mode switch
    attr_accessor :test_mode

    # Set's the default value's to nil and false
    # @return [Hash] configuration options
    def init!
      @defaults = {
        :@partner_id => nil,
        :@profile_key => nil,
        :@report_url => nil,
        :@return_url => nil,
        :@test_mode => false
      }
    end

    # Resets the value's to there previous value (instance_variable)
    # @return [Hash] configuration options
    def reset!
      @defaults.each { |key, value| instance_variable_set(key, value) }
    end

    # Set's the new value's as instance variables
    # @return [Hash] configuration options
    def update!
      @defaults.each do |key, value|
        instance_variable_set(key, value) unless instance_variable_defined?(key)
      end
    end
  end
  init!
  reset!
end


lib/ideal-mollie/engine.rb
module IdealMollie
  #
  # Simpel extend on the +Rails::Engine+ to add support for a new config section within
  # the environment configs
  #
  # @example default
  #   # /config/environments/development.rb
  #   config.ideal_mollie.partner_id = 123456
  #   config.ideal_mollie.report_url = "http://example.org/report"
  #   config.ideal_mollie.return_url = "http://example.org/return"
  #   config.ideal_mollie.test_mode = false
  #
  # @example optional profile_key when having to deal with multiple profile's
  #   # /config/environments/development.rb
  #   config.ideal_mollie.partner_id = 123456
  #   config.ideal_mollie.report_url = "http://example.org/report"
  #   config.ideal_mollie.return_url = "http://example.org/return"
  #   config.ideal_mollie.test_mode = false
  #   config.ideal_mollie.profile_key = "123abc45"
  #
  class Engine < Rails::Engine
    config.ideal_mollie = IdealMollie::Config
  end
end


lib/ideal-mollie/order_result.rb
module IdealMollie
  #
  # Object representing a "OrderResult" object with attributes provided by Mollie
  #
  # @example
  #   Order.new({
  #     :transaction_id => "4b99662febb42ce6f889d9c57f5cf3fa",
  #     :amount => 1465,
  #     :currency => "EUR",
  #     :payed => true,
  #     :consumer => {
  #       :consumerName => "Hr J Janssen",
  #       :consumerAccount => "P001234567",
  #       :consumerCity => "Amsterdam"
  #     },
  #     :message => "This iDEAL-order has successfuly been payed for,
  #                  and this is the first time you check it.",
  #     :status => "Expired"
  #   })
  class OrderResult < Order
    # @return [Boolean] Order was paid.
    attr_accessor :paid
    # @return [String] The name of the customer.
    attr_accessor :customer_name
    # @return [String] The bankaccount number of the customer.
    attr_accessor :customer_account
    # @return [String] The city of the customer.
    attr_accessor :customer_city
    # @return [String] Status of the Order.
    attr_accessor :status

    #
    # Initializer to transform a +Hash+ into an OrderResult object
    #
    # @param [Hash] values
    def initialize(values=nil)
      return if values.nil?

      super(values)

      @paid = values["payed"] == "true" ? true : false if values.has_key?("payed")
      @status = values["status"].to_s if values.has_key?("status")

      if values.has_key?("consumer")
        consumer = values["consumer"]
        @customer_name = consumer["consumerName"].to_s if consumer.has_key?("consumerName")
        @customer_account = consumer["consumerAccount"].to_s if consumer.has_key?("consumerAccount")
        @customer_city = consumer["consumerCity"].to_s if consumer.has_key?("consumerCity")
      end
    end
  end
end


lib/ideal-mollie.rb
# Gems
require "faraday"
require "faraday_middleware"
require "multi_xml"

# Files
require "ideal-mollie/bank"
require "ideal-mollie/config"
require "ideal-mollie/engine" if defined?(Rails) && Rails::VERSION::MAJOR.to_i >= 3
require "ideal-mollie/ideal_exception"
require "ideal-mollie/order"
require "ideal-mollie/order_result"

#
# IdealMollie Module
#
module IdealMollie
  # Mollie API url
  MOLLIE_URL = "https://secure.mollie.nl/xml/ideal"

  #
  # List of supported banks.
  #
  # @visibility public
  #
  # @example
  #   IdealMollie.banks
  #
  # @return [Array<IdealMollie::Bank>] list of supported +Bank+'s.
  def self.banks
    response = IdealMollie.request("banklist")

    banks = response["bank"]
    banks = [banks] unless banks.is_a?Array # to array if it isn't already

    banks.inject([]) { |result, bank| result << IdealMollie::Bank.new(bank); result }
  end

  #
  # Create a new order.
  #
  # @visibility public
  #
  # @param [Hash or int] hash_or_amount The amount of money to transfer (defined in cents) or a Hash.
  # @param [String] description The description of the payment on the bank transfer.
  # @param [String] bank_id The id of the bank selected from one of the supported banks.
  # @param [String] return_url Optional override of the return url specified in the Config
  #
  # @example
  #   # as a Hash
  #   IdealMollie.new_order(amount: 1000, description: "Ordernumber #123: new gadget", bank_id: "0031")
  #   IdealMollie.new_order(
  #     amount: 1000,
  #     description: "Ordernumber #123: new gadget",
  #     bank_id: "0031",
  #     return_url: "http://override.url/controller/return_action",
  #     report_url: "http://override.url/controller/report_action"
  #   )
  #
  #   # as arguments
  #   IdealMollie.new_order(1000, "Ordernumber #123: new gadget", "0031")
  #   IdealMollie.new_order(1000, "Ordernumber #123: new gadget", "0031", "http://override.url/controller/return_action")
  #   IdealMollie.new_order(1000, "Ordernumber #123: new gadget", "0031", "http://override.url/controller/return_action", "http://override.url/controller/report_action")
  #
  # @return [IdealMollie::Order] the +Order+.
  def self.new_order(hash_or_amount, description=nil, bank_id=nil, return_url=nil, report_url=nil)
    amount = hash_or_amount
    if hash_or_amount.is_a?(Hash)
      amount = hash_or_amount[:amount]
      description = hash_or_amount[:description]
      bank_id = hash_or_amount[:bank_id]
      return_url = hash_or_amount[:return_url] if hash_or_amount.has_key?(:return_url)
      report_url = hash_or_amount[:report_url] if hash_or_amount.has_key?(:report_url)
    end
    params = new_order_params(amount, description, bank_id, return_url, report_url)
    response = IdealMollie.request("fetch", params)

    IdealMollie::Order.new(response["order"])
  end

  #
  # Check the status of a order.
  #
  # @visibility public
  #
  # @param [String] transaction_id the transaction to verify.
  #
  # @example
  #   IdealMollie.check_order("4b99662febb42ce6f889d9c57f5cf3fa")
  #
  # @note Once a transaction is paid, only the next time you verify the
  #   transaction will the value of +paid+ be +true+.
  #   Else it will be +false+.
  #
  # @return [IdealMollie::OrderResult] the +OrderResult+.
  def self.check_order(transaction_id)
    response = IdealMollie.request("check", {
      :partnerid => Config.partner_id,
      :transaction_id => transaction_id
    })
    IdealMollie::OrderResult.new(response["order"])
  end

  class << self
    MultiXml.parser = :nokogiri

    #
    # Perform a request to the https xlm server of mollie and return the responded xml
    # which is converted to an hash
    #
    # @param [String] action The action to perform.
    # @param [Hash] params Additional parameters to send like partner_id
    #
    # @raise [IdealMollie::IdealException] When a error is returned by Mollie
    #
    # @return [Hash] the returned XML as a +Hash+
    def request(action, params={})
      params.merge!(:a => action)
      params.merge!(:testmode => Config.test_mode)
      request = connection.post do |req|
        req.url("", params)
      end
      response = request.body["response"]

      if response.has_key?("item")
        error = response["item"]
        raise IdealMollie::IdealException.new(error["errorcode"], error["message"], error["type"])
      end
      response
    end

    #
    # Builds a +Hash+ with the parameters, that are needed for making a new order
    # Makes sure the +return_url+ is set to the correct value
    #
    # @param [int] amount The amount of money to transfer (defined in cents).
    # @param [String] description The description of the payment on the bank transfer.
    # @param [String] bank_id The id of the bank selected from one of the supported banks.
    # @param [String] return_url Optional override of the return url specified in the Config
    #
    # @return [Hash] the parameter +Hash+ for the new order.
    def new_order_params(amount, description, bank_id, return_url=nil, report_url=nil)
      return_url = Config.return_url if return_url.nil?
      report_url = Config.report_url if report_url.nil?

      params = {
        :partnerid => Config.partner_id,
        :reporturl => report_url,
        :returnurl => return_url,
        :description => description,
        :amount => amount,
        :bank_id => bank_id
      }

      unless Config.profile_key.nil?
        params.merge!({:profile_key => Config.profile_key})
      end

      params
    end

    private

    def connection
      @connection ||= Faraday::Connection.new(:url => IdealMollie::MOLLIE_URL,
                                  :headers => {:user_agent => "Ruby-IdealMollie"},
                                  :ssl => {:verify => false}) do |builder|
        builder.adapter Faraday.default_adapter
        builder.use FaradayMiddleware::ParseXml
      end
    end
  end
end
